{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Introduction to VAPr: package for the aggregation of genomic variant data \n",
    "\n",
    "#### Author: C. Mazzaferro & Kathleen Fisch\n",
    "#### Email: cmazzafe@ucsd.edu\n",
    "#### Date: June 2016\n",
    " \n",
    "## Outline of Notebook\n",
    "<a id = \"toc\"></a>\n",
    "1. <a href = \"#background\">Background</a>\n",
    "2. <a href = \"#setup\"> Annovar Set-Up</a>\n",
    "  * <a href = \"#naming\">Set paths and names</a>\n",
    "3. <a href = \"#mongo\">Set Up MongoDB</a>\n",
    "  * <a href = \"#parse\">Parse to MongoDB</a>\n",
    "4. <a href = \"#export\">Export CSV and VCF files</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id = \"background\"></a>\n",
    "## Background\n",
    "\n",
    "This notebook will walk you through the basic steps of how variants coming from a VCF can be annotated efficiently and thoroughly using the package VAPr. In particular, the package is aimed at providing a way of retrieving variant information using [ANNOVAR](http://annovar.openbioinformatics.org/en/latest/) and [myvariant.info](myvariant.info) and consolidating it in conveninent formats. It is well-suited for bioinformaticians interested in aggregating variant information into a single database for ease of use and to provide higher analysis capabities.\n",
    "\n",
    "For a more complete description of the functionalities of the package, you can visit the [VAPr Sample Usage Notebook](https://github.com/ucsd-ccbb/VAPr/blob/master/VAPr%20Sample%20Usage.ipynb) and/or check the full documentation on GitHub. \n",
    "\n",
    "\n",
    "*Note to reviewers*: this notebook is in all its functionalities the same as the one in the official documentation (https://github.com/ucsd-ccbb/VAPr/blob/master/Quick%20Start-Up%20Guide.ipynb). We have added some notes here regarding performance and regarding the file structure of this instance."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id = \"setup\"></a>\n",
    "## Annovar Set-Up\n",
    "\n",
    "Annovar* is one of the most popular open source variant annotation packages currently available. It's a collection of comand line perl scripts that levarage on up-to-date publicly available datasets, and allows the execution of region, gene, and filter based annotations. VAPr employs the functionalities of Annovar by providing wrapper funcions that execute automatically the database downloads and update, as well as the annotation itself. In particular, the wrapper was written in order to provide the user with the latest annotation data, but also minizing the overhead of  having to learn a new toolkit fro scratch. The wrapper this takes care of:\n",
    "\n",
    " 1. Building a syntactically correct command line string command\n",
    " 2. Storing downloaded files, as well as annotated ones to their appropriate locations\n",
    " 3. Automatic updates of databases\n",
    " \n",
    "The databases currently supported as default are the following:\n",
    " \n",
    "- knownGene\n",
    "- tfbsConsSites\n",
    "- cytoBand\n",
    "- genomicSuperDups\n",
    "- esp6500siv2_all\n",
    "- 1000g2015aug_all\n",
    "- popfreq_all\n",
    "- clinvar_20140929\n",
    "- cosmic70\n",
    "- nci60\n",
    "\n",
    "Which can be expanded or restricted by the user, depending on his or her research needs. \n",
    "\n",
    "Required set up for this step: download annovar from [here](http://www.openbioinformatics.org/annovar/annovar_download_form.php) and extract it to the location you'd like the databases to live in. The entire disk size of the databases will be around 25 GB, so make sure you have such space available.\n",
    " \n",
    "\n",
    "*[ANNOVAR: Functional annotation of genetic variants from next-generation sequencing data Nucleic Acids Research](http://nar.oxfordjournals.org/content/38/16/e164) , 38:e164, 2010\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id = \"naming\"></a>\n",
    "### Set paths and names\n",
    "Import modules and, set the variable paths and initialize the object `sub_process`, that contains the functions in charge of dealing with annovar\n",
    "\n",
    "*Note to reviewers*\n",
    "\n",
    "Two files are available for testing in this envirnoment: \n",
    "\n",
    " 1. test_file_one.vcf (approx 25000 variants)\n",
    " 2. test_file_two.vcf (approx 250000 variants).\n",
    " \n",
    "The workflow focuses more on the first file, since running the entire pipeline for file two will require ~1.5 hours. For the first file, it shouldn't take more than a few minutes (running time doesn't increase linearly with the number of variants). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# third party modules\n",
    "import pandas as pd\n",
    "import importlib\n",
    "\n",
    "# variantannotation modules\n",
    "from VAPr import parser_models, annovar_suprocess, file_writer\n",
    "\n",
    "\n",
    "IN_PATH = \"/data/vcf_files/test_file_one.vcf\"\n",
    "OUT_PATH = \"/data/csv_files/\"\n",
    "ANNOVAR_PATH = \"/data/annovar/\"   #Folder of the extrscted annovar package \n",
    "sub_process = annovar_suprocess.AnnovarWrapper(IN_PATH, OUT_PATH, ANNOVAR_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants in vcf file: 26625\n"
     ]
    }
   ],
   "source": [
    "#Get an estimate of the number of variants in the vcf file\n",
    "print(\"Number of variants in vcf file: %i\" % sum(1 for line in open(IN_PATH)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently downloading database file: hg19_knownGene\n",
      "Currently downloading database file: hg19_kgXref\n",
      "Currently downloading database file: hg19_knownGeneMrna\n",
      "Currently downloading database file: hg19_knownGene\n",
      "\n",
      "Annovar finished dowloading on file : hg19_knownGene. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: tfbsConsSites\n",
      "Currently downloading database file: tfbsConsSites\n",
      "\n",
      "Annovar finished dowloading on file : tfbsConsSites. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: cytoBand\n",
      "Currently downloading database file: cytoBand\n",
      "\n",
      "Annovar finished dowloading on file : cytoBand. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: targetScanS\n",
      "Currently downloading database file: targetScanS\n",
      "\n",
      "Annovar finished dowloading on file : targetScanS. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: genomicSuperDups\n",
      "Currently downloading database file: genomicSuperDups\n",
      "\n",
      "Annovar finished dowloading on file : genomicSuperDups. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: hg19_esp6500siv2_all\n",
      "Currently downloading database file: hg19_esp6500siv2_all\n",
      "Currently downloading database file: hg19_esp6500siv2_all\n",
      "\n",
      "Annovar finished dowloading on file : hg19_esp6500siv2_all. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: hg19_1000g2015aug\n",
      "Currently downloading database file: hg19_esp6500siv2_all\n",
      "Currently downloading database file: hg19_AFR\n",
      "\n",
      "Annovar finished dowloading on file : hg19_AFR. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: hg19_popfreq_all_20150413\n",
      "Currently downloading database file: hg19_ALL\n",
      "\n",
      "Annovar finished dowloading on file : hg19_ALL. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: hg19_clinvar_20161128\n",
      "Currently downloading database file: hg19_clinvar_20161128\n",
      "Currently downloading database file: hg19_clinvar_20161128\n",
      "\n",
      "Annovar finished dowloading on file : hg19_clinvar_20161128. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: hg19_cosmic70\n",
      "Currently downloading database file: hg19_cosmic70\n",
      "Currently downloading database file: hg19_popfreq_all_20150413\n",
      "Currently downloading database file: hg19_cosmic70\n",
      "\n",
      "Annovar finished dowloading on file : hg19_cosmic70. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n",
      "Currently downloading database file: hg19_nci60\n",
      "Currently downloading database file: hg19_nci60\n",
      "Currently downloading database file: hg19_cosmic70\n",
      "Currently downloading database file: hg19_nci60\n",
      "\n",
      "Annovar finished dowloading on file : hg19_nci60. A .txt file has been created in the ANNOVAR_PATH directory\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Finished downloading databases to /data/annovar/humandb/'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_process.download_dbs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently working on VCF file: test_file_one_annotated, field variant_function\n",
      "Currently working on VCF file: test_file_one_annotated, field exonic_variant_function\n",
      "Currently working on VCF file: test_file_one_annotated, field variant_function\n",
      "Currently working on VCF file: test_file_one_annotated, field exonic_variant_function\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_tfbsConsSites\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_tfbsConsSites\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_cytoBand\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_cytoBand\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_targetScanS\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_targetScanS\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_genomicSuperDupsCurrently working on VCF file: test_file_one_annotated, field hg19_genomicSuperDups\n",
      "\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_esp6500siv2_all_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_esp6500siv2_all_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_esp6500siv2_all_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_esp6500siv2_all_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field 2015_08_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field 2015_08_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field 2015_08_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field 2015_08_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_popfreq_all_20150413_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_popfreq_all_20150413_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_popfreq_all_20150413_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_popfreq_all_20150413_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_clinvar_20161128_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_clinvar_20161128_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_clinvar_20161128_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_clinvar_20161128_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_cosmic70_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_cosmic70_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_cosmic70_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_cosmic70_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_nci60_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_nci60_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_nci60_filtered\n",
      "Currently working on VCF file: test_file_one_annotated, field hg19_nci60_dropped\n",
      "Currently working on VCF file: test_file_one_annotated, field txt\n",
      "\n",
      "Annovar finished working on file : test_file_one_annotated. A .txt file has been created in the OUT_PATH directory\n",
      "\n",
      "Currently working on VCF file: test_file_one_annotated, field txt\n",
      "\n",
      "Annovar finished working on file : test_file_one_annotated. A .txt file has been created in the OUT_PATH directory\n",
      "\n",
      "CPU times: user 1.24 s, sys: 260 ms, total: 1.5 s\n",
      "Wall time: 1min 45s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Finished running ANNOVAR on /data/vcf_files/test_file_one.vcf'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "sub_process.run_annovar()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "~1 minute for 25k variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id = \"mongo\"></a>\n",
    "## Set up MongoDB \n",
    "\n",
    "*Note to reviewers*: MongoDB is already installed. Running `sudo service mongod restart` should get it back up and running. \n",
    "\n",
    "As mentioned, a MongoDB instance must be installed and running. Platform-specific installation instructions can be found [here](https://www.mongodb.com/download-center?jmp=docs&_ga=1.190648363.1890535995.1486057213#community), alongside with download of the latest distributions.\n",
    "\n",
    "![Imgur](http://i.imgur.com/FjLUDFb.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A great introduction to the concepts relevant to MongoDB, and specifically how to interact with it using python can be found [here](https://docs.mongodb.com/getting-started/python/introduction/). Of particular interest is the explanation on how documents are formatted and stored inside a the database. Variants will follow that format as well: a sample entry of the database will look roughly as follows (the document for this variant (HGVS_id: chr20:g.25194768A>G) has been reduced to half of its actual size). \n",
    "\n",
    "\n",
    "```python\n",
    "{\n",
    "  \"_id\": ObjectId(\"5887c6d3bc644e51c028971a\"),\n",
    "  \"chr\": \"chr19\",\n",
    "  \"cytoband\": {\n",
    "    \"Region\": \"13\",\n",
    "    \"Sub_Band\": \"41\",\n",
    "    \"Chromosome\": 19,\n",
    "    \"Band\": \"q\",\n",
    "    \"Name\": \"19q13.41\"\n",
    "  },\n",
    "  \"alt\": \"C\",\n",
    "  \"hg19\": {\n",
    "    \"end\": 25194768,\n",
    "    \"start\": 25194768\n",
    "  },\n",
    "  \n",
    "  ...\n",
    "  \n",
    "  \"otherinfo\": [\n",
    "    \"GT:AD:DP:GQ:PL\",\n",
    "    \"1/1:0,2:2:6:89,6,0\"\n",
    "  ],\n",
    "  \"hgvs_id\": \"chr20:g.25194768A>G\",\n",
    "  \"esp6500siv2_all\": \"0.71\",\n",
    "  \"ref\": \"T\",\n",
    "  \"chrom\": \"20\",\n",
    "  \"grasp\": {\n",
    "    \"exclusively_male_female\": \"n\",\n",
    "    \"initial_sample_description\": [\n",
    "      \"Up to 46186 EA individuals\",\n",
    "      [\n",
    "        \"Up to 3445 EA cases\",\n",
    "        \" 6935 EA controls\"\n",
    "      ],\n",
    "      \"69395 EA individuals\",\n",
    "      \"69395 EA individuals\"\n",
    "    ],\n",
    "    \"creation_date\": \"8/17/12\",\n",
    "    \"gwas_ancestry_description\": \"European\",\n",
    "    \"srsid\": 6083780,\n",
    "    \"platform_snps_passing_qc\": [\n",
    "      \"Affymetrix & Illumina [~2.5 million] (imputed)\",\n",
    "      \"Illumina [528745]\",\n",
    "      [\n",
    "        \"Affymetrix\",\n",
    "        \" Illumina & Perlegen [~2.5 million] (imputed)\"\n",
    "      ],\n",
    "      [\n",
    "        \"Affymetrix\",\n",
    "        \" Illumina & Perlegen [~2.5 million] (imputed)\"\n",
    "      ]\n",
    "    ],\n",
    "    \"hg19\": {\n",
    "      \"chr\": 20,\n",
    "      \"pos\": 25194768\n",
    "    },\n",
    "    \"publication\": [\n",
    "      {\n",
    "        \"pmid\": 20081858,\n",
    "        \"phenotype\": \"HOMA-B\",\n",
    "        \"p_value\": 0.033930000000000001825,\n",
    "        \"snpid\": \"rs6083780\",\n",
    "        \"paper_phenotype_description\": [\n",
    "          \"Glucose homeostasis traits (fasting glucose\",\n",
    "          \" fasting insulin\",\n",
    "          \" HOMA-B\",\n",
    "          \" HOMA-IR)\"\n",
    "        ],\n",
    "        \"date_pub\": \"1/17/2010\",\n",
    "        \"title\": \"New genetic loci implicated in fasting glucose homeostasis and their impact on type 2 diabetes risk.\",\n",
    "        \"location_within_paper\": \"FullData\",\n",
    "        \"paper_phenotype_categories\": \"Quantitative trait(s);Type 2 diabetes (T2D);Blood-related\",\n",
    "        \"journal\": \"Nat Genet\"\n",
    "      },\n",
    "      {\n",
    "        \"pmid\": 20522523,\n",
    "        \"phenotype\": \"Partial epilepsy\",\n",
    "        \"p_value\": 0.022849999999999998784,\n",
    "        \"snpid\": \"rs6083780\",\n",
    "        \"paper_phenotype_description\": \"Epilepsy (partial epilepsy)\",\n",
    "        \"date_pub\": \"6/22/2010\",\n",
    "        \"title\": \"Common genetic variation and susceptibility to partial epilepsies: a genome-wide association study.\",\n",
    "        \"location_within_paper\": \"FullScan\",\n",
    "        \"paper_phenotype_categories\": \"Neuro;Epilepsy\",\n",
    "        \"journal\": \"Brain\"\n",
    "      }\n",
    "    ],\n",
    "    \"last_curation_date\": \"8/17/12\",\n",
    "    \"replication\": [\n",
    "      {\n",
    "        \"total_samples\": 76558,\n",
    "        \"european\": 76558\n",
    "      },\n",
    "      {\n",
    "        \"total_samples\": 133661,\n",
    "        \"european\": 133661\n",
    "      }\n",
    "    ],\n",
    "    \"discovery\": [\n",
    "      {\n",
    "        \"total_samples\": 46186,\n",
    "        \"european\": 46186\n",
    "      },\n",
    "      {\n",
    "        \"total_samples\": 10380,\n",
    "        \"european\": 10380\n",
    "      }\n",
    "    ],\n",
    "    \"hupfield\": \"Jan2014\",\n",
    "    \"includes_male_female_only_analyses\": \"n\",\n",
    "    \"in_gene\": \"(ENTPD6)\",\n",
    "    \"replication_sample_description\": [\n",
    "      \"up to 76558 EA individuals\",\n",
    "      \"NR\",\n",
    "      \"133661 EA individuals\",\n",
    "      \"133661 EA individuals\"\n",
    "    ]\n",
    "  },\n",
    "  \"start\": 51447065\n",
    "} ```\n",
    "\n",
    "\n",
    "The richness of the data derives from the usage of MyVariant.info services and the high-availability of the datasets hosted by them. Further, being updated at least monthly, their databases can be guaranteed to deliver the most accurate and relevat data for specific variants. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id = \"parse\"></a>\n",
    "### Prepare to parse data to MongoDB\n",
    "\n",
    "*to initiate a mongo db instance (MongoDB is already install in this tutorial), run `sudo service mongod restart`\n",
    "\n",
    "Here, we finally parse the data to Mongo. Assuming all the installation steps have been taken, and mongodb is running (the command `sudo chkconfig mongod on` should return [ OK ]. Check [this](https://docs.mongodb.com/manual/tutorial/install-mongodb-on-amazon/) for more information on how to perform and check the installation on an AWS instance. \n",
    "\n",
    "#### File names and variables\n",
    "\n",
    "The required file names and variables are the following:\n",
    "\n",
    "\n",
    "- CSV annotated (assuming its been annotated with ANNOVAR already. If not, refer to the documentation for instructions on how to perform annotation).\n",
    "- VCF (original file containing the variants\n",
    "- MongoDB collection and Database names. A mongoDB instance must be running upon calling the main VAPr function, but the parsing will be done automatically by the underlying `pymongo` library.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "csv_file = '/data/csv_files/test_file_one_annotated.hg19_multianno.txt'\n",
    "vcf_file = '/data/vcf_files/test_file_one.vcf'\n",
    "\n",
    "collection_name = 'My_Variant_Collection_File_One'\n",
    "db_name = 'My_Variant_Database'\n",
    "\n",
    "pars = parser_models.VariantParsing(vcf_file, collection_name, db_name, annotated_file=csv_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-602...done.\n",
      "querying 1-604...done.\n",
      "querying 1-605...done.\n",
      "querying 1-602...done.\n",
      "querying 1-601...done.\n",
      "querying 1-604...done.\n",
      "querying 1-603...done.\n",
      "querying 1-601...done.\n",
      "querying 1-603...done.\n",
      "querying 1-601...done.\n",
      "querying 1-604...done.\n",
      "querying 1-604...done.\n",
      "querying 1-606...done.\n",
      "querying 1-601...done.\n",
      "querying 1-608...done.\n",
      "querying 1-606...done.\n",
      "querying 1-604...done.\n",
      "querying 1-603...done.\n",
      "querying 1-602...done.\n",
      "querying 1-602...done.\n",
      "querying 1-604...done.\n",
      "querying 1-605...done.\n",
      "querying 1-600...done.\n",
      "querying 1-604...done.\n",
      "querying 1-605...done.\n",
      "querying 1-600...done.\n",
      "querying 1-605...done.\n",
      "querying 1-603...done.\n",
      "querying 1-601...done.\n",
      "querying 1-600...done.\n",
      "querying 1-606...done.\n",
      "querying 1-604...done.\n",
      "querying 1-608...done.\n",
      "querying 1-600...done.\n",
      "querying 1-604...done.\n",
      "querying 1-603...done.\n",
      "querying 1-605...done.\n",
      "querying 1-603...done.\n",
      "querying 1-604...done.\n",
      "querying 1-603...done.\n",
      "querying 1-601...done.\n",
      "querying 1-607...done.\n",
      "querying 1-605...done.\n",
      "querying 1-605...done.\n",
      "querying 1-101...done.\n",
      "CPU times: user 1min 12s, sys: 0 ns, total: 1min 12s\n",
      "Wall time: 2min 53s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Done'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time \n",
    "pars.push_to_db() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "~3 minutes for 25k variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Implement a filter\n",
    "\n",
    " - filter 1: ThousandGenomeAll < 0.05 or info not available\n",
    " - filter 2: ESP6500siv2_all < 0.05 or info not available\n",
    " - filter 3: cosmic70 information is present\n",
    " - filter 4: Func_knownGene is exonic, splicing, or both\n",
    " - filter 5: ExonicFunc_knownGene is not \"synonymous SNV\"\n",
    " - filter 6: Read Depth (DP) > 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#names and paths \n",
    "rare_cancer_variants_csv = \"/data/out_files/filtered_csv.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variants found that match rarity criteria: 1\n"
     ]
    }
   ],
   "source": [
    "#Apply filter.\n",
    "from VAPr import MongoDB_querying, file_writer\n",
    "\n",
    "filter_collection = MongoDB_querying.Filters(db_name, collection_name)\n",
    "rare_cancer_variants = filter_collection.rare_cancer_variant()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inspect entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1000g2015aug_all': 0.00159744,\n",
       " '_id': ObjectId('589e35893c5b990f8441ed78'),\n",
       " 'alt': 'G',\n",
       " 'cadd': {'1000g': {'af': 0.5, 'afr': 0.5, 'amr': 0.5, 'asn': 0.5, 'eur': 0.5},\n",
       "  '_license': 'http://goo.gl/bkpNhq',\n",
       "  'alt': 'G',\n",
       "  'anc': 'A',\n",
       "  'annotype': 'CodingTranscript',\n",
       "  'bstatistic': 872,\n",
       "  'chmm': {'bivflnk': 0.0,\n",
       "   'enh': 0.0,\n",
       "   'enhbiv': 0.008,\n",
       "   'het': 0.016,\n",
       "   'quies': 0.504,\n",
       "   'reprpc': 0.087,\n",
       "   'reprpcwk': 0.354,\n",
       "   'tssa': 0.0,\n",
       "   'tssaflnk': 0.0,\n",
       "   'tssbiv': 0.008,\n",
       "   'tx': 0.0,\n",
       "   'txflnk': 0.008,\n",
       "   'txwk': 0.008,\n",
       "   'znfrpts': 0.0},\n",
       "  'chrom': 7,\n",
       "  'consdetail': 'missense',\n",
       "  'consequence': 'NON_SYNONYMOUS',\n",
       "  'consscore': 7,\n",
       "  'cpg': 0.03,\n",
       "  'dna': {'helt': 0.03, 'mgw': 0.4, 'prot': 3.05, 'roll': 3.04},\n",
       "  'encode': {'exp': 40.15,\n",
       "   'h3k27ac': 7.04,\n",
       "   'h3k4me1': 3.0,\n",
       "   'h3k4me3': 3.0,\n",
       "   'nucleo': 3.7},\n",
       "  'exon': '4/5',\n",
       "  'fitcons': 0.527649,\n",
       "  'gc': 0.55,\n",
       "  'gene': {'ccds_id': 'CCDS5872.1',\n",
       "   'cds': {'cdna_pos': 514,\n",
       "    'cds_pos': 508,\n",
       "    'rel_cdna_pos': 0.64,\n",
       "    'rel_cds_pos': 0.68},\n",
       "   'feature_id': 'ENST00000311737',\n",
       "   'gene_id': 'ENSG00000204983',\n",
       "   'genename': 'PRSS1',\n",
       "   'prot': {'domain': 'ndomain', 'protpos': 170, 'rel_prot_pos': 0.69}},\n",
       "  'gerp': {'n': 3.28, 'rs': 149.7, 'rs_pval': 8.08864e-16, 's': -6.56},\n",
       "  'grantham': 56,\n",
       "  'isderived': 'TRUE',\n",
       "  'isknownvariant': 'TRUE',\n",
       "  'istv': 'FALSE',\n",
       "  'length': 0,\n",
       "  'mapability': {'20bp': 1, '35bp': 1},\n",
       "  'min_dist_tse': 476,\n",
       "  'min_dist_tss': 1374,\n",
       "  'mutindex': -26,\n",
       "  'naa': 'E',\n",
       "  'oaa': 'K',\n",
       "  'phast_cons': {'mammalian': 0.0, 'primate': 0.011, 'vertebrate': 0.0},\n",
       "  'phred': 0.001,\n",
       "  'phylop': {'mammalian': -1.491, 'primate': -2.595, 'vertebrate': -1.594},\n",
       "  'polyphen': {'cat': 'benign', 'val': 0},\n",
       "  'pos': 142460335,\n",
       "  'rawscore': -2.903274,\n",
       "  'ref': 'A',\n",
       "  'scoresegdup': 0.93,\n",
       "  'segway': 'R5',\n",
       "  'sift': {'cat': 'tolerated', 'val': 0.93},\n",
       "  'type': 'SNV'},\n",
       " 'chr': 'chr7',\n",
       " 'chrom': '7',\n",
       " 'cosmic': {'alt': 'G',\n",
       "  'chrom': '7',\n",
       "  'cosmic_id': 'COSM42958',\n",
       "  'hg19': {'end': 142460335, 'start': 142460335},\n",
       "  'mut_freq': 0.26,\n",
       "  'mut_nt': 'A>G',\n",
       "  'ref': 'A',\n",
       "  'tumor_site': 'central_nervous_system'},\n",
       " 'cosmic70': 'ID=COSM42958;OCCURENCE=1(central_nervous_system)',\n",
       " 'cytoband': {'Band': 'q', 'Chromosome': 7, 'Name': '7q34', 'Region': '34'},\n",
       " 'dbnsfp': {'1000gp3': {'ac': 8,\n",
       "   'af': 0.001597444089456869,\n",
       "   'afr_ac': 0,\n",
       "   'afr_af': 0.0,\n",
       "   'amr_ac': 0,\n",
       "   'amr_af': 0.0,\n",
       "   'eas_ac': 0,\n",
       "   'eas_af': 0.0,\n",
       "   'eur_ac': 0,\n",
       "   'eur_af': 0.0,\n",
       "   'sas_ac': 8,\n",
       "   'sas_af': 0.0081799591002045},\n",
       "  'aa': {'alt': 'E',\n",
       "   'codon_degeneracy': 0,\n",
       "   'codonpos': 1,\n",
       "   'pos': ['184', '160', '170', '170', '120'],\n",
       "   'ref': 'K',\n",
       "   'refcodon': 'AAG'},\n",
       "  'alt': 'G',\n",
       "  'ancestral_allele': 'A',\n",
       "  'cds_strand': '+',\n",
       "  'chrom': '7',\n",
       "  'dann': {'_license': 'https://goo.gl/IeLhCq',\n",
       "   'rankscore': 0.00193,\n",
       "   'score': 0.11901653757417319},\n",
       "  'eigen': {'phred': 0.004467639,\n",
       "   'raw': -2.14012544445484,\n",
       "   'raw_rankscore': 0.00104},\n",
       "  'eigen-pc': {'raw': -2.21358174570785, 'raw_rankscore': 0.00102},\n",
       "  'ensembl': {'geneid': 'ENSG00000204983',\n",
       "   'proteinid': ['ENSP00000417854',\n",
       "    'ENSP00000481361',\n",
       "    'ENSP00000308720',\n",
       "    'ENSP00000479959',\n",
       "    'ENSP00000419912'],\n",
       "   'transcriptid': ['ENST00000486171',\n",
       "    'ENST00000619214',\n",
       "    'ENST00000311737',\n",
       "    'ENST00000612126',\n",
       "    'ENST00000492062']},\n",
       "  'exac': {'ac': 60686,\n",
       "   'adj_ac': 60685,\n",
       "   'adj_af': 0.5,\n",
       "   'af': 0.4998,\n",
       "   'afr_ac': 5195,\n",
       "   'afr_af': 0.5,\n",
       "   'amr_ac': 5789,\n",
       "   'amr_af': 0.5,\n",
       "   'eas_ac': 4326,\n",
       "   'eas_af': 0.5,\n",
       "   'fin_ac': 3306,\n",
       "   'fin_af': 0.5,\n",
       "   'nfe_ac': 33363,\n",
       "   'nfe_af': 0.5,\n",
       "   'sas_ac': 8252,\n",
       "   'sas_af': 0.5},\n",
       "  'exac_nonpsych': {'ac': 45357,\n",
       "   'adj_ac': 45356,\n",
       "   'adj_af': 0.5,\n",
       "   'af': 0.4998,\n",
       "   'afr_ac': 5186,\n",
       "   'afr_af': 0.5,\n",
       "   'amr_ac': 5783,\n",
       "   'amr_af': 0.5,\n",
       "   'eas_ac': 2818,\n",
       "   'eas_af': 0.5,\n",
       "   'fin_ac': 1901,\n",
       "   'fin_af': 0.5,\n",
       "   'nfe_ac': 21064,\n",
       "   'nfe_af': 0.5},\n",
       "  'exac_nontcga': {'ac': 53090,\n",
       "   'adj_ac': 53089,\n",
       "   'adj_af': 0.5,\n",
       "   'af': 0.4999,\n",
       "   'afr_ac': 4525,\n",
       "   'afr_af': 0.5,\n",
       "   'amr_ac': 5608,\n",
       "   'amr_af': 0.5,\n",
       "   'eas_ac': 3932,\n",
       "   'eas_af': 0.5,\n",
       "   'fin_ac': 3306,\n",
       "   'fin_af': 0.5,\n",
       "   'nfe_ac': 27171,\n",
       "   'nfe_af': 0.5,\n",
       "   'sas_ac': 8200,\n",
       "   'sas_af': 0.5},\n",
       "  'fathmm': {'pred': ['D', '.', 'D', '.', 'D'],\n",
       "   'rankscore': 0.88235,\n",
       "   'score': [-2.37, None, -2.37, None, -2.37]},\n",
       "  'fathmm-mkl': {'coding_group': 'AEFBI',\n",
       "   'coding_pred': 'N',\n",
       "   'coding_rankscore': 0.01084,\n",
       "   'coding_score': 0.00183},\n",
       "  'genename': 'PRSS1',\n",
       "  'genocanyon': {'rankscore': 0.17816, 'score': 0.176891010153521},\n",
       "  'gerp++': {'nr': 3.28, 'rs': -6.56, 'rs_rankscore': 0.01672},\n",
       "  'gm12878': {'confidence_value': 2,\n",
       "   'fitcons_rankscore': 0.05779,\n",
       "   'fitcons_score': 0.457222},\n",
       "  'h1-hesc': {'confidence_value': 0,\n",
       "   'fitcons_rankscore': 0.1464,\n",
       "   'fitcons_score': 0.547309},\n",
       "  'hg18': {'end': 142139909, 'start': 142139909},\n",
       "  'hg19': {'end': 142460335, 'start': 142460335},\n",
       "  'hg38': {'end': 142752484, 'start': 142752484},\n",
       "  'huvec': {'confidence_value': 0,\n",
       "   'fitcons_rankscore': 0.10106,\n",
       "   'fitcons_score': 0.530356},\n",
       "  'integrated': {'confidence_value': 0,\n",
       "   'fitcons_rankscore': 0.08333,\n",
       "   'fitcons_score': 0.446893},\n",
       "  'interpro_domain': 'Peptidase S1, PA clan|Serine proteases, trypsin domain',\n",
       "  'lrt': {'converted_rankscore': 0.05542,\n",
       "   'omega': 1.17538,\n",
       "   'pred': 'N',\n",
       "   'score': 0.582927},\n",
       "  'metalr': {'pred': 'T', 'rankscore': 0.00011, 'score': 0.0},\n",
       "  'metasvm': {'pred': 'T', 'rankscore': 0.07107, 'score': -1.0812},\n",
       "  'mutationassessor': {'pred': 'N', 'rankscore': 0.02885, 'score': -0.42},\n",
       "  'mutationtaster': {'AAE': ['K184E', 'K170E'],\n",
       "   'converted_rankscore': 0.08979,\n",
       "   'model': ['simple_aae', 'simple_aae'],\n",
       "   'pred': ['N', 'N'],\n",
       "   'score': [1, 1]},\n",
       "  'phastcons': {'100way': {'vertebrate': 0.0, 'vertebrate_rankscore': 0.06329},\n",
       "   '20way': {'mammalian': 0.0, 'mammalian_rankscore': 0.01567}},\n",
       "  'phylo': {'p100way': {'vertebrate': -3.971, 'vertebrate_rankscore': 0.00362},\n",
       "   'p20way': {'mammalian': -5.639, 'mammalian_rankscore': 8e-05}},\n",
       "  'polyphen2': {'_license': 'http://goo.gl/6Cz4Ae',\n",
       "   'hdiv': {'pred': 'B', 'rankscore': 0.02634, 'score': 0.0},\n",
       "   'hvar': {'pred': 'B', 'rankscore': 0.01257, 'score': 0.0}},\n",
       "  'provean': {'pred': ['N', '.', 'N', '.', 'N'],\n",
       "   'rankscore': 0.01556,\n",
       "   'score': [0.93, None, 0.93, None, 1.03]},\n",
       "  'ref': 'A',\n",
       "  'reliability_index': 10,\n",
       "  'rsid': 'rs201550522',\n",
       "  'sift': {'converted_rankscore': 0.1311,\n",
       "   'pred': ['T', '.', 'T', '.', 'T'],\n",
       "   'score': [0.387, None, 0.378, None, 0.33]},\n",
       "  'siphy_29way': {'logodds': 5.5807,\n",
       "   'logodds_rankscore': 0.16418,\n",
       "   'pi': {'a': 0.163, 'c': 0.5271, 'g': 0.2031, 't': 0.1068}},\n",
       "  'uniprot': [{'acc': 'E7EQ64', 'pos': '184'},\n",
       "   {'acc': 'P07477', 'pos': '170'}],\n",
       "  'vest3': {'_license': 'http://goo.gl/jTko4F',\n",
       "   'rankscore': ['0.213', '0.181', '0.181'],\n",
       "   'score': [0.213, 0.181, 0.181],\n",
       "   'transcriptid': ['ENST00000492062', 'NM_002769.4', 'ENST00000486171'],\n",
       "   'transcriptvar': ['K120E', 'K170E', 'K184E']}},\n",
       " 'dbsnp': {'allele_origin': 'unspecified',\n",
       "  'alleles': [{'allele': 'A', 'freq': 0.9984},\n",
       "   {'allele': 'G', 'freq': 0.001597}],\n",
       "  'alt': 'G',\n",
       "  'chrom': '7',\n",
       "  'class': 'SNV',\n",
       "  'dbsnp_build': 137,\n",
       "  'flags': ['ASP', 'G5', 'KGPhase1', 'KGPhase3', 'NSM', 'REF', 'S3D', 'SLO'],\n",
       "  'gene': {'geneid': '5644', 'symbol': 'PRSS1'},\n",
       "  'gmaf': 0.001597,\n",
       "  'hg19': {'end': 142460335, 'start': 142460335},\n",
       "  'ref': 'A',\n",
       "  'rsid': 'rs201550522',\n",
       "  'validated': True,\n",
       "  'var_subtype': 'ts',\n",
       "  'vartype': 'snp'},\n",
       " 'end': 142460335,\n",
       " 'exac': {'ac': {'ac': 60686,\n",
       "   'ac_adj': 60685,\n",
       "   'ac_afr': 5195,\n",
       "   'ac_amr': 5789,\n",
       "   'ac_eas': 4326,\n",
       "   'ac_female': 27047,\n",
       "   'ac_fin': 3306,\n",
       "   'ac_het': 60683,\n",
       "   'ac_hom': 1,\n",
       "   'ac_male': 33638,\n",
       "   'ac_nfe': 33363,\n",
       "   'ac_oth': 454,\n",
       "   'ac_sas': 8252},\n",
       "  'af': 0.5,\n",
       "  'alleles': 'G',\n",
       "  'alt': 'G',\n",
       "  'an': {'an': 121410,\n",
       "   'an_adj': 121372,\n",
       "   'an_afr': 10390,\n",
       "   'an_amr': 11578,\n",
       "   'an_eas': 8652,\n",
       "   'an_female': 54094,\n",
       "   'an_fin': 6612,\n",
       "   'an_male': 67278,\n",
       "   'an_nfe': 66728,\n",
       "   'an_oth': 908,\n",
       "   'an_sas': 16504},\n",
       "  'baseqranksum': 5.55,\n",
       "  'chrom': 7,\n",
       "  'clippingranksum': -0.526,\n",
       "  'culprit': 'MQRankSum',\n",
       "  'dp': 17954823,\n",
       "  'filter': 'InbreedingCoeff_Filter',\n",
       "  'fs': 5.029,\n",
       "  'het': {'ac_het': 60683,\n",
       "   'het_afr': 5195,\n",
       "   'het_amr': 5789,\n",
       "   'het_eas': 4326,\n",
       "   'het_fin': 3306,\n",
       "   'het_nfe': 33363,\n",
       "   'het_oth': 454,\n",
       "   'het_sas': 8250},\n",
       "  'hom': {'ac_hom': 1,\n",
       "   'hom_afr': 0,\n",
       "   'hom_amr': 0,\n",
       "   'hom_eas': 0,\n",
       "   'hom_fin': 0,\n",
       "   'hom_nfe': 0,\n",
       "   'hom_oth': 0,\n",
       "   'hom_sas': 1},\n",
       "  'inbreedingcoeff': -0.9991,\n",
       "  'mq': {'mq': 52.89, 'mq0': 0, 'mqranksum': -8.792},\n",
       "  'ncc': 7,\n",
       "  'pos': 142460335,\n",
       "  'qd': 11.49,\n",
       "  'qual': 303853780.6,\n",
       "  'readposranksum': 0.516,\n",
       "  'ref': 'A',\n",
       "  'type': 'snp',\n",
       "  'vqslod': 93.46},\n",
       " 'exac_nontcga': {'ac': {'ac': 53090,\n",
       "   'ac_adj': 53089,\n",
       "   'ac_afr': 4525,\n",
       "   'ac_amr': 5608,\n",
       "   'ac_eas': 3932,\n",
       "   'ac_female': 22927,\n",
       "   'ac_fin': 3306,\n",
       "   'ac_het': 53087,\n",
       "   'ac_hom': 1,\n",
       "   'ac_male': 30162,\n",
       "   'ac_nfe': 27171,\n",
       "   'ac_oth': 347,\n",
       "   'ac_sas': 8200},\n",
       "  'af': 0.5,\n",
       "  'alleles': 'G',\n",
       "  'alt': 'G',\n",
       "  'an': {'an': 106208,\n",
       "   'an_adj': 106180,\n",
       "   'an_afr': 9050,\n",
       "   'an_amr': 11216,\n",
       "   'an_eas': 7864,\n",
       "   'an_female': 45854,\n",
       "   'an_fin': 6612,\n",
       "   'an_male': 60326,\n",
       "   'an_nfe': 54344,\n",
       "   'an_oth': 694,\n",
       "   'an_sas': 16400},\n",
       "  'baseqranksum': 5.55,\n",
       "  'chrom': 7,\n",
       "  'clippingranksum': -0.526,\n",
       "  'culprit': 'MQRankSum',\n",
       "  'fs': 5.029,\n",
       "  'het': {'het_afr': 4525,\n",
       "   'het_amr': 5608,\n",
       "   'het_eas': 3932,\n",
       "   'het_fin': 3306,\n",
       "   'het_nfe': 27171,\n",
       "   'het_oth': 347,\n",
       "   'het_sas': 8198},\n",
       "  'hom': {'hom_afr': 0,\n",
       "   'hom_amr': 0,\n",
       "   'hom_eas': 0,\n",
       "   'hom_fin': 0,\n",
       "   'hom_nfe': 0,\n",
       "   'hom_oth': 0,\n",
       "   'hom_sas': 1},\n",
       "  'inbreedingcoeff': -0.9991,\n",
       "  'mq': {'mq': 52.89, 'mq0': 0, 'mqranksum': -8.792},\n",
       "  'ncc': 7,\n",
       "  'pos': 142460335,\n",
       "  'qd': 11.49,\n",
       "  'readposranksum': 0.516,\n",
       "  'ref': 'A',\n",
       "  'type': 'snp',\n",
       "  'vqslod': 93.46},\n",
       " 'exonicfunc_knowngene': 'nonsynonymous SNV',\n",
       " 'func_knowngene': 'exonic',\n",
       " 'gene_knowngene': 'PRSS1',\n",
       " 'geno2mp': {'hpo_count': 62},\n",
       " 'genomicsuperdups': {'Name': 'chr9:33776725', 'Score': 0.93194},\n",
       " 'genotype': {'alleles': [250, 194],\n",
       "  'filter_passing_reads_count': [444],\n",
       "  'genotype': '0/1',\n",
       "  'genotype_lieklihoods': [6676.0, 0.0, 9732.0]},\n",
       " 'hg19': {'end': 142460335, 'start': 142460335},\n",
       " 'hgvs_id': 'chr7:g.142460335A>G',\n",
       " 'mutdb': {'alt': 'G',\n",
       "  'chrom': '7',\n",
       "  'cosmic_id': '42958',\n",
       "  'hg19': {'end': 142460335, 'start': 142460335},\n",
       "  'mutpred_score': 0.329,\n",
       "  'ref': 'A',\n",
       "  'rsid': None,\n",
       "  'strand': 'p'},\n",
       " 'nci60': 0.48,\n",
       " 'otherinfo': ['GT:AD:DP:GQ:PL', '0/1:250,194:444:99:6676,0,9732'],\n",
       " 'ref': 'A',\n",
       " 'snpeff': {'ann': {'cdna': {'length': '817', 'position': '525'},\n",
       "   'cds': {'length': '744', 'position': '508'},\n",
       "   'effect': 'missense_variant',\n",
       "   'feature_id': 'NM_002769.4',\n",
       "   'feature_type': 'transcript',\n",
       "   'gene_id': 'PRSS1',\n",
       "   'genename': 'PRSS1',\n",
       "   'hgvs_c': 'c.508A>G',\n",
       "   'hgvs_p': 'p.Lys170Glu',\n",
       "   'protein': {'length': '247', 'position': '170'},\n",
       "   'putative_impact': 'MODERATE',\n",
       "   'rank': '4',\n",
       "   'total': '5',\n",
       "   'transcript_biotype': 'Coding'}},\n",
       " 'start': 142460335,\n",
       " 'tfbsconssites': {'Name': 'V$TCF11MAFG_01', 'Score': 758.0},\n",
       " 'vcf': {'alt': 'G', 'position': '142460335', 'ref': 'A'},\n",
       " 'wellderly': {'adviser_score': '5~PRSS1~Common, Predicted Neutral',\n",
       "  'allele_aa': 'E',\n",
       "  'alleles': [{'allele': 'A', 'freq': 0.9025},\n",
       "   {'allele': 'G', 'freq': 0.0975}],\n",
       "  'alt': 'G',\n",
       "  'chrom': '7',\n",
       "  'coding_impact': 'Nonsynonymous',\n",
       "  'gene': ['TCRBV5S1A1T', 'TCRBV2S1', 'TCRVB', 'BV03S1J2.2', 'PRSS1', 'PRSS2'],\n",
       "  'genotypes': [{'count': 39, 'freq': 0.195, 'genotype': 'A/G'},\n",
       "   {'count': 161, 'freq': 0.8049999999999999, 'genotype': 'A/A'}],\n",
       "  'hg19': {'end': 142460335, 'start': 142460335},\n",
       "  'original_aa': 'K',\n",
       "  'polyphen': 'benign',\n",
       "  'pos': 142460335,\n",
       "  'protein_pos': ['170', '110'],\n",
       "  'ref': 'A',\n",
       "  'vartype': 'snp'}}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rare_cancer_variants[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other filters\n",
    "The filter may have been too 'aggressive' or there simply aren't enough variants that match the criteria. In any case, it is possible to implement manually a filter in order to obtain more results. The query syntax is the one implemented by the library [pymongo](http://api.mongodb.com/python/current/tutorial.html), developed by MongoDB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pymongo import MongoClient\n",
    "\n",
    "client = MongoClient()\n",
    "db = getattr(client, db_name)\n",
    "collection = getattr(db, collection_name)\n",
    "\n",
    "\n",
    "filtered = collection.find({\"$and\": [\n",
    "                                   {\"$or\": [{\"esp6500siv2_all\": {\"$lt\": 0.1}}, {\"esp6500siv2_all\": {\"$exists\": False}}]},\n",
    "                                   {\"$or\": [{\"func_knowngene\": \"exonic\"}, {\"func_knowngene\": \"splicing\"}]},\n",
    "                                   {\"genotype.filter_passing_reads_count\": {\"$gte\": 1}},\n",
    "                                   {\"cosmic70\": {\"$exists\": True}},\n",
    "                                   {\"1000g2015aug_all\": {\"$exists\": True}}\n",
    "                         ]})\n",
    "\n",
    "as_list = list(filtered)\n",
    "len(as_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Or a dataframe for easy manipulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['1000g2015aug_all', '_id', 'alt', 'cadd', 'chr', 'chrom', 'cosmic',\n",
       "       'cosmic70', 'cytoband', 'dbnsfp', 'dbsnp', 'end', 'exac',\n",
       "       'exac_nontcga', 'exonicfunc_knowngene', 'func_knowngene',\n",
       "       'gene_knowngene', 'geno2mp', 'genomicsuperdups', 'genotype', 'hg19',\n",
       "       'hgvs_id', 'mutdb', 'nci60', 'otherinfo', 'ref', 'snpeff', 'start',\n",
       "       'tfbsconssites', 'vcf', 'wellderly'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(as_list)\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1000g2015aug_all</th>\n",
       "      <th>_id</th>\n",
       "      <th>alt</th>\n",
       "      <th>cadd</th>\n",
       "      <th>chr</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cosmic</th>\n",
       "      <th>cosmic70</th>\n",
       "      <th>cytoband</th>\n",
       "      <th>dbnsfp</th>\n",
       "      <th>...</th>\n",
       "      <th>hgvs_id</th>\n",
       "      <th>mutdb</th>\n",
       "      <th>nci60</th>\n",
       "      <th>otherinfo</th>\n",
       "      <th>ref</th>\n",
       "      <th>snpeff</th>\n",
       "      <th>start</th>\n",
       "      <th>tfbsconssites</th>\n",
       "      <th>vcf</th>\n",
       "      <th>wellderly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.393970</td>\n",
       "      <td>589e35563c5b990f8441c699</td>\n",
       "      <td>A</td>\n",
       "      <td>{'cpg': 0.24, 'consdetail': ['synonymous', 're...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>{'alt': 'T', 'chrom': '1', 'ref': 'C', 'mut_fr...</td>\n",
       "      <td>ID=COSM1320072;OCCURENCE=2(thyroid),1(ovary)</td>\n",
       "      <td>{'Band': 'p', 'Chromosome': 1, 'Sub_Band': '2'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:g.120612006G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22</td>\n",
       "      <td>[GT:AD:DP:GQ:PL, 0/1:9,8:17:99:240,0,335]</td>\n",
       "      <td>G</td>\n",
       "      <td>{'ann': [{'feature_type': 'transcript', 'rank'...</td>\n",
       "      <td>120612006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'alt': 'A', 'position': '120612006', 'ref': 'G'}</td>\n",
       "      <td>{'adviser_score': '5~NOTCH2~Common, Predicted ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.289137</td>\n",
       "      <td>589e35593c5b990f8441c96e</td>\n",
       "      <td>AGACCATGGCCCCGCCCAGTCCCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=COSM1745478;OCCURENCE=4(urinary_tract)</td>\n",
       "      <td>{'Band': 'q', 'Chromosome': 1, 'Sub_Band': '1'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:g.203186950_203186951insAGACCATGGCCCCGCCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GT:AD:DP:GQ:PL, 1/1:0,3:3:9:135,9,0]</td>\n",
       "      <td>-</td>\n",
       "      <td>{'lof': {'gene_id': 'CHIT1', 'genename': 'CHIT...</td>\n",
       "      <td>203186950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'alt': 'CAGACCATGGCCCCGCCCAGTCCCT', 'position...</td>\n",
       "      <td>{'adviser_score': '5~CHIT1~Common, Predicted N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.706669</td>\n",
       "      <td>589e35593c5b990f8441ca33</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=COSM244564;OCCURENCE=1(NS),2(pancreas),2(la...</td>\n",
       "      <td>{'Band': 'q', 'Chromosome': 1, 'Name': '1q43',...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1:g.240255569_240255571del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GT:AD:DP:GQ:PL, 1/1:0,2:2:6:80,6,0]</td>\n",
       "      <td>GGC</td>\n",
       "      <td>{'ann': {'feature_type': 'transcript', 'rank':...</td>\n",
       "      <td>240255569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'alt': 'G', 'position': '240255568', 'ref': '...</td>\n",
       "      <td>{'alt': 'G', 'chrom': '1', 'gene': 'FMN2', 're...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.062899</td>\n",
       "      <td>589e355e3c5b990f8441ce6a</td>\n",
       "      <td>G</td>\n",
       "      <td>{'cpg': 0.05, 'consdetail': 'intron', 'gene': ...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=COSM3836683;OCCURENCE=1(breast)</td>\n",
       "      <td>{'Band': 'q', 'Chromosome': 2, 'Sub_Band': '2'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chr2:g.120015164A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GT:AD:DP:GQ:PL, 0/1:1,4:5:30:156,0,30]</td>\n",
       "      <td>A</td>\n",
       "      <td>{'ann': [{'feature_id': 'NM_182915.2', 'gene_i...</td>\n",
       "      <td>120015164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'alt': 'G', 'position': '120015164', 'ref': 'A'}</td>\n",
       "      <td>{'alt': 'G', 'chrom': '2', 'gene': 'STEAP3', '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   1000g2015aug_all                       _id                       alt  \\\n",
       "0          0.393970  589e35563c5b990f8441c699                         A   \n",
       "1          0.289137  589e35593c5b990f8441c96e  AGACCATGGCCCCGCCCAGTCCCT   \n",
       "2          0.706669  589e35593c5b990f8441ca33                         -   \n",
       "3          0.062899  589e355e3c5b990f8441ce6a                         G   \n",
       "\n",
       "                                                cadd   chr chrom  \\\n",
       "0  {'cpg': 0.24, 'consdetail': ['synonymous', 're...  chr1     1   \n",
       "1                                                NaN  chr1     1   \n",
       "2                                                NaN  chr1     1   \n",
       "3  {'cpg': 0.05, 'consdetail': 'intron', 'gene': ...  chr2     2   \n",
       "\n",
       "                                              cosmic  \\\n",
       "0  {'alt': 'T', 'chrom': '1', 'ref': 'C', 'mut_fr...   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "\n",
       "                                            cosmic70  \\\n",
       "0       ID=COSM1320072;OCCURENCE=2(thyroid),1(ovary)   \n",
       "1          ID=COSM1745478;OCCURENCE=4(urinary_tract)   \n",
       "2  ID=COSM244564;OCCURENCE=1(NS),2(pancreas),2(la...   \n",
       "3                 ID=COSM3836683;OCCURENCE=1(breast)   \n",
       "\n",
       "                                            cytoband dbnsfp  \\\n",
       "0  {'Band': 'p', 'Chromosome': 1, 'Sub_Band': '2'...    NaN   \n",
       "1  {'Band': 'q', 'Chromosome': 1, 'Sub_Band': '1'...    NaN   \n",
       "2  {'Band': 'q', 'Chromosome': 1, 'Name': '1q43',...    NaN   \n",
       "3  {'Band': 'q', 'Chromosome': 2, 'Sub_Band': '2'...    NaN   \n",
       "\n",
       "                         ...                          \\\n",
       "0                        ...                           \n",
       "1                        ...                           \n",
       "2                        ...                           \n",
       "3                        ...                           \n",
       "\n",
       "                                             hgvs_id  mutdb nci60  \\\n",
       "0                                chr1:g.120612006G>A    NaN  0.22   \n",
       "1  chr1:g.203186950_203186951insAGACCATGGCCCCGCCC...    NaN   NaN   \n",
       "2                      chr1:g.240255569_240255571del    NaN   NaN   \n",
       "3                                chr2:g.120015164A>G    NaN   NaN   \n",
       "\n",
       "                                   otherinfo  ref  \\\n",
       "0  [GT:AD:DP:GQ:PL, 0/1:9,8:17:99:240,0,335]    G   \n",
       "1      [GT:AD:DP:GQ:PL, 1/1:0,3:3:9:135,9,0]    -   \n",
       "2       [GT:AD:DP:GQ:PL, 1/1:0,2:2:6:80,6,0]  GGC   \n",
       "3    [GT:AD:DP:GQ:PL, 0/1:1,4:5:30:156,0,30]    A   \n",
       "\n",
       "                                              snpeff      start tfbsconssites  \\\n",
       "0  {'ann': [{'feature_type': 'transcript', 'rank'...  120612006           NaN   \n",
       "1  {'lof': {'gene_id': 'CHIT1', 'genename': 'CHIT...  203186950           NaN   \n",
       "2  {'ann': {'feature_type': 'transcript', 'rank':...  240255569           NaN   \n",
       "3  {'ann': [{'feature_id': 'NM_182915.2', 'gene_i...  120015164           NaN   \n",
       "\n",
       "                                                 vcf  \\\n",
       "0  {'alt': 'A', 'position': '120612006', 'ref': 'G'}   \n",
       "1  {'alt': 'CAGACCATGGCCCCGCCCAGTCCCT', 'position...   \n",
       "2  {'alt': 'G', 'position': '240255568', 'ref': '...   \n",
       "3  {'alt': 'G', 'position': '120015164', 'ref': 'A'}   \n",
       "\n",
       "                                           wellderly  \n",
       "0  {'adviser_score': '5~NOTCH2~Common, Predicted ...  \n",
       "1  {'adviser_score': '5~CHIT1~Common, Predicted N...  \n",
       "2  {'alt': 'G', 'chrom': '1', 'gene': 'FMN2', 're...  \n",
       "3  {'alt': 'G', 'chrom': '2', 'gene': 'STEAP3', '...  \n",
       "\n",
       "[4 rows x 31 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export\n",
    "Files can also be exported to CSV, as well as VCF files while mantaining the annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Finished writing annotated, filtered CSV file'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Crete writer object for filtered lists:\n",
    "my_writer = file_writer.FileWriter(db_name, collection_name)\n",
    "\n",
    "rare_cancer_variants_csv = '/data/out_files/rare_vars.csv'\n",
    "\n",
    "#cancer variants filtered files\n",
    "my_writer.generate_annotated_csv(rare_cancer_variants, rare_cancer_variants_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export Filtered (VCF) Files\n",
    "This is possible as well, although a bit trickier. Re-creating a vcf file from a list of variants requires an index file from the original file from which a filtered version will be built. This can be achieved using the tabix tool from http://genometoolbox.blogspot.com/2013/11/installing-tabix-on-unix.html.\n",
    "\n",
    "*NOTE*: The annotations from MyVariant.info will be added to the INFO field\n",
    "\n",
    "*NOTE to reviewers*: Tabix is already installed in this instance\n",
    "\n",
    "**Step 1**: Download Tabix:\n",
    "\n",
    "`tar xvjf tabix-0.2.6.tar.bz2`  \n",
    "`cd tabix-0.2.6`    \n",
    "`make`    \n",
    "`export PATH=$PATH:/path_to_tabix/tabix-0.2.6`   \n",
    "\n",
    "**Step 2**: create a .vcf.gz file from original vcf:\n",
    "\n",
    "`bgzip -c file.vcf > file.vcf.gz`\n",
    "\n",
    "\n",
    "**Step 3**: run tabix on the input vcf file:\n",
    "\n",
    "`tabix -p vcf file.vcf.gz`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Specify name of files and path "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Finished writing annotated, filtered VCF file'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rare_cancer_variants_vcf = \"/data/out_files/rare_variants.vcf\"\n",
    "input_vcf_compressed = \"/data/vcf_files/test_file_one.vcf.gz\"\n",
    "\n",
    "filtered_variant_list = list(filtered)\n",
    "\n",
    "#Crete writer object for filtered lists:\n",
    "my_writer = file_writer.FileWriter(db_name, collection_name)\n",
    "my_writer.generate_annotated_vcf(filtered_variant_list, input_vcf_compressed, rare_cancer_variants_vcf)\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [py35]",
   "language": "python",
   "name": "Python [py35]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
